### PR TITLE
fix diff_pairs_mask to fit original conditions

### PR DESCRIPTION
**Pull request by DavidMetzIMT**
Hy,

I found I think a solution to my [issue-1211346947](https://github.com/liubenyuan/pyEIT/issues/41#issue-1211346947)

the `self.J` was empty because of `voltage_meter` was corrupted. The bool array mask was wrong... (@ChabaneArmaury I must admit I the not at the begining of the if condition is confusing...) 

>not (m == drv_a or m == drv_b or n == drv_a or n == drv_b) or meas_current

is equivalent to 

>diff_pairs_mask = np.array(
        ((m != drv_a) & (m != drv_b) & (n != drv_a) & (n != drv_b)) | meas_current
    )

I didn't have time to check `voltage_meter_nd` but I will do it and after testing it I will make PR if needed!

Cheers

### CODE REVIEW

**Comment by DavidMetzIMT** [comment]
Hy,
Here some details of the changes  I pulled:
### In `fem.py`

- fix the bool array mask in the methods `voltage_meter` and `voltage_meter_nd` which gived out an empty `diff_pairs` for `parser = 'meas_current'`
 
the original masking condition was confusing:

> not (m == drv_a or m == drv_b or n == drv_a or n == drv_b) or meas_current

here a less confusing that I implemented (De Morgan simplification):
>  (m != drv_a and m != drv_b and n != drv_a and n != drv_b) or meas_current 

- fix/enhanced the docstrings parser can be str or list[str]

Notes:
@liubenyuan: It would be good to clean `fem.py` and maybe only keep the optimized computation of @ChabaneAmaury (that way you could delete the `vector` arg and keep the code simple)

### In `unittest_fem.py`
-  first implement tests for both methods `voltage_meter` and `voltage_meter_nd` for `parser = None` and for `parser = 'meas_current'`  (sry I'm not used to write unittest... yes bad habbits...)

Those tests asset if the `diff_pairs` are not empty, but a more enhanced tests should be made...

### In `base.py`

- To profit of the optimized computation during new instantiation of jac, bp and greit solver (which all are based on a first calculation of the jacobian matrix though the call of the improved `solve_eit` method), I add the `vector` (Default to `False`) argument to the `EitBase` constructor in order to be able to use improved calculation by passing it during initial calculation of the jacobian matrix though the call of the improved `solve_eit` method at line 91.  That argument is memory under a new attr `self.vector` (e.g. for gn calclation in jac or for greit)

### In `jac.py`

- in `gn` method use the att `self.vector` if not passed (to avoid multiple default value)

### In `greit.py`

- in `setup` method the jacobian is recalculated for generated grid by calling the `_build_dist` method where the `self.vector` attr is now passed to use improved calculation

Note: @liubenyuan: for greit two jacobian are calculated one by instantiation (which is not used after) and one after setup (the one which is finally used). Maybe it would be good for all solvers to solve fwd during setup (when we get all parameters) and not during instantiation and to avoid to build a potentionally wrong `self.H` matrix ... . I though to add for all solver a generic `h_matrix` method (like in jac )which is than call at the end of each setup. what do you think? i you want I can have a look on that and pull some changes.

I hope didn't forget something important!
Many thanks for the great job on PyEIT!

David

**Comment by liubenyuan** [comment]
That's a good idea. So you must run setup() after initializing each solver.

**Comment by liubenyuan** [comment]
Follow up: currently only `fem.py` uses `voltage_meter`, and the `diff_op` generated by voltage_meter is only used by subtract_row operates on f and jac. I think maybe these two function can be merged?

File: pyeit/eit/fem.py [hunk 1]
```diff
@@ -474,7 +468,7 @@ def voltage_meter(ex_line, n_el=16, step=1, parser=None) -> np.ndarray:
     #         # the order of m, n matters
     #         v.append([n, m])
     # diff_pairs = np.array(v)
-    return diff_pairs
+    return arr[diff_pairs_mask]
```
Comment (liubenyuan) [inline pyeit/eit/fem.py:471]: https://github.com/liubenyuan/pyEIT/pull/42#issuecomment-1106638940

**Review by liubenyuan — COMMENTED** [review]
Could you please clean up the voltage_meter_nd code? We could drop the non-vectorized code and keep only _nd (rename voltage_meter_nd to voltage_meter, etc) code if these code are well documented and easy to understand.

File: pyeit/eit/fem.py [hunk 2]
```diff
@@ -537,23 +531,21 @@ def voltage_meter_nd(ex_mat, n_el=16, step=1, parser=None):
...
                 )
                 | meas_current
             )
             for i in range(m.shape[0])
         ]
     )
     arr = np.array([np.array([n[i], m[i]]).T for i in range(n.shape[0])])
-    diff_pairs = np.array(
-        [arr[i, ~np.array((diff_pairs_mask[i]))] for i in range(arr.shape[0])]
+    return np.array(
```
Comment (liubenyuan) [inline pyeit/eit/fem.py:545]: the returns of voltage_meter and voltage_meter_nd should not be convert to ndarray? Please see the warning message.

File: pyeit/eit/fem.py [hunk 3]
```diff
@@ -537,23 +531,21 @@ def voltage_meter_nd(ex_mat, n_el=16, step=1, parser=None):
...
-                    (m[i] == drv_a[i])
-                    | (m[i] == drv_b[i])
-                    | (n[i] == drv_a[i])
-                    | (n[i] == drv_b[i])
+                    (m[i] != drv_a[i])
+                    & (m[i] != drv_b[i])
+                    & (n[i] != drv_a[i])
+                    & (n[i] != drv_b[i])
                 )
                 | meas_current
```
Comment (liubenyuan) [inline pyeit/eit/fem.py:539]: shouldn't we do `if meas_current: return arr, else: diff_mask=[], return arr[diff_mask]`?

File: pyeit/eit/fem.py [hunk 4]
```diff
@@ -537,23 +531,21 @@ def voltage_meter_nd(ex_mat, n_el=16, step=1, parser=None):
         [
             (
                 (
-                    (m[i] == drv_a[i])
-                    | (m[i] == drv_b[i])
-                    | (n[i] == drv_a[i])
-                    | (n[i] == drv_b[i])
+                    (m[i] != drv_a[i])
```
Comment (liubenyuan) [inline pyeit/eit/fem.py:534]: these lines can be shortened. drv_a[i] and drv_b[i] are from a line of ex_mat, which previous called ex_line, so in essense, we are testing whether mi or ni are in ex_line, and then exclude the corresponding diff_paris.

I write a test script in doing this, though the `np.delete` and `np.isin` seems not efficient.
```python
    arr = [np.array([ni, mi]).T for ni, mi in zip(n, m)]
    if not meas_current:
        arr = [np.delete(a, np.any(np.isin(a, ex_mat[i]), axis=1), axis=0) for i, a in enumerate(arr)]

    return np.array(arr)
```

**Review by liubenyuan — COMMENTED** [review]
Please see the notes in `voltage_meter_nd`. And sure, do include only the vectorization code, and try to keep the codes and comments clean and easy to understand.

**Comment by DavidMetzIMT** [comment]
Hy,
you were very productive during the weekend! :)
I will try to clean up the code today or tomorrow! and take in account your comments!
Notes: this will also include deleting the `vector` argument through whole package!
David

**Comment by DavidMetzIMT** [comment]
Hy,

I merged the vectorized calculation in `fem.py`
- all *_nd methods / function are now standards
- remove `vector` argument from all modules (assure back compatibility on public methods and avoid TypeError: * got an unexpected keyword argument 'vector', I add **kwargs to those methods (`gn` of JAC and in EItBase constructor)
- I improved: 
   - voltage_meter using only matrices  >> 35% less computation time
   - subtract_row using only matrices  >> 33% less computation time
   - smear using only matrices, however I couldn't see a significant improvement! oder computation is still availbe and can be selected used the `new` flag

Tests:
I didn't have time to make a correct unit test but I tested the changes using some of your examples scripts.
did you have time to make one?

Remarks: I enhances the doctsrings  and clean them by the way the [Google Style Docstrings](https://google.github.io/styleguide/pyguide.html#s3.8.1-comments-in-doc-strings) (see [here](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy) for an overview) is much more compact in the code! if you ar aggree I will in next PR fromta them!.

**Pull request merged**